### DISCUSSION TITLE
[Change in Behaviour]  Incorrect content download options shown - appears to be a browser tab identification issue etc (ie multiple compounding issues)

### DISCUSSION THREAD

**Discussion by BundyBob** [aclap-dev/video-downloadhelper#179] [Help and Q&A]
### Your question

ENVIRONMENT INFORMATION
--------------------------------
Video DownloadHelper
Version 8.2.0.23 (stable)
Browser locale: en-US
Production build
Built on 2024-02-14
Build options: browser=firefox, linuxlic=false, noyt=false, target=mozilla

Found companion app:  VdhCoApp 2.0.19
Companion app binary:  C:\Program Files\DownloadHelper CoApp\vdhcoapp.exe

Firefox Browser 122.0.1 (64-bit) - portable edition

Processor:	12th Gen Intel(R) Core(TM) i9-12900H   2.50 GHz
Installed RAM:	64.0 GB (63.7 GB usable)
System type:	64-bit operating system, x64-based processor
Operating System:  Microsoft Windows 11 Pro
OS Version:  10.0.22631 Build 22631

------
ISSUE
------
When I right-mouse on a tab with media content available, I expect the context menu to show me the available download streaming configurations available for that page.  Since the last automated update of the Video DownloadHelper, the available list has not reflected the browser tab I was on (I normally have many tabs open when I'm working - some but not all have streaming media sources available).  Today when I needed to download a stream, I noticed the single reference related to a different tab.  I tried shifting that tab to the right of the tab I was on to see if the displayed left-right tab order was the issue.  It was not (ie the displayed info still related to the other tab).  I pulled that tab out to form a separate window but that did not resolve the issue.  Finally I closed that tab to see if the list would populate with content from another tab.  It didn't - there were no files listed in the context menu (or from the extension icon menu on the browser.  

I reordered tabs to place my desired tab to the tab position 3rd from the left before closing the browser (the first 2 tabs have no media streaming directly available on them i.e. you would need to click a link to change the page to access such media) and then terminated the Firefox session (allowing for a restore session option to be available).  

On relaunching the browser and restoring the session, I went to the 3rd tab and selected the context menu for downloading the stream.  The label above the media download options for the stream correctly identified the text between the HTML <title></title> identifiers for the tab.  I downloaded the first stream option being for the 1920x1080 size.

I then went to the next 7 tabs to the right & each one showed the <title></title> field value for that tab, HOWEVER when I went back to the original starting point (3rd tab from the left) and check the context list of available downloads, the media label no longer corresponded to the current tab but that of the rightmost tab I had been to.  I then repeated access to the 7 tabs to the right of the starting point and each time context list showed the media label as being of the rightmost tab of the group of tabs I had accessed.  There were more tabs to the right that I hadn't accessed.

I next accessed the tab immediately to the right of the previous rightmost tab I had tested.  It displayed the correct <title></title> identifier for that tab.  Going back to the starting point, the context menu showed the label for that last right-most tab accessed & likewise cycling through the other tabs to the right up to the last right-most tab.  

The other quirk (i.e. bug) noted was that some of the streaming options listed under the inconsistent labeling, showed a play button icon rather than the expected download icon.  Using a tab where I hadn't attempted a download I clicked the play button.  It triggered the launch of a player with a previously downloaded file (e.g. from the starting tab for this sequence of testing in the current Firefox session).  Once this had happened, the next time I tried the context menu, the label no longer displayed the <title></title>for the rightmost tab but rather the <title></title>for the 3rd tab from the left which also matched the content that played when previously testing the play button option from the context menu from the rightmost tab used.

So there would seem to be a need for a revision of the code controlling this area for the current Video DownloadHelper version as identified at the top of this issue.  In this test the media download options were listed prefixed with "DASH streaming" (but issues that occurred after the last update but prior to me testing to identify an issue, were not necessarily of this same type

* The media download options should only be limited to the currently selected tab otherwise it would be an unmanageable situation.
* The label for the media option is not currently guaranteed to reflect the choices for the tab you are on.
* If you delete the tab for the content referred to in the media label, you may find that you no longer have media listed for download
* Where multiple content labels exist in the context menu, you are not guaranteed that it will list media download options for the tab you are on.
* Files that are downloaded appear to be short of the original content's length (3+ seconds short of the correct size).
* Media options listed under a content label (context menu) that has a play button, MAY NOT correctly reference a file that contains the content relating to that label, nor may it reflect the streaming properties associated with that media option line (e.g. where media line indicates a window size of 1280 x 720 for a streaming rate of 275Kbps, the file actually played can have totally different characteristics).
* For the tab you are on, if there is an option to start a download but the label reflects content from a different tab, don't expect the resultant file to reflect the content for the tab you are on - it is more likely to be a file for the tab reflected in the label (but unconfirmed as guaranteed without further specific testing of a range of varied scenarios).
* Information specific for the web page contained on a Firefox tab appears to be overwritten by data relating to a different tab.  It smells like data being written/read from a class rather than an instance of a class that I would expect i.e. the Video DownloadHelper code should treat and store data separately for each tab and any change to that data should relate only to the current web page displayed for that tab.

I don't think the above is an exhaustive list of the problems - other commitments limit spending further time on this issue at this stage.

### Addon & CoApp version and details

Video DownloadHelper
Version 8.2.0.23 (stable)
Browser locale: en-US
Production build
Built on 2024-02-14
Build options: browser=firefox, linuxlic=false, noyt=false, target=mozilla

Found companion app:  VdhCoApp 2.0.19
Companion app binary:  C:\Program Files\DownloadHelper CoApp\vdhcoapp.exe

Firefox Browser 122.0.1 (64-bit) - portable edition

Processor:	12th Gen Intel(R) Core(TM) i9-12900H   2.50 GHz
Installed RAM:	64.0 GB (63.7 GB usable)
System type:	64-bit operating system, x64-based processor
Operating System:  Microsoft Windows 11 Pro
OS Version:  10.0.22631 Build 22631

**Comment by GrampaWildWilly** [comment]
I don't know what the devs are going to say (I'm not one of them, I'm just another user) but I found your problem description terribly confusing.  I would probably follow it better if you gave screenshots showing each of the steps of what you did.  URLs interspersed between the screenshots would probably help your cause as well.

**Comment by BundyBob** [comment]
An image file relating to the testing.  The web site shouldn't be relevant but it is behind a login for paid account so I won't be passing on credentials.  I imagine that it will access a video stream sourced from Vimeo.

![v8-2-0-23=DoesNotAccessCorrectTab=#A](https://github.com/aclap-dev/video-downloadhelper/assets/111411485/76ca7696-9bf0-4892-b3e6-ed04e4964d3f)

**Reply by GrampaWildWilly** [reply]
You're quite right about not posting the URLs behind the paywall, & sharing logon credentials is generally considered a sin around here anyway so that was never on the cards.

I think the thing that is the biggest problem is that the title of the download being proposed sometimes matches the content of the current tab & sometimes does not.  It seems the closing of other tabs seems to influence this.  It's too bad one of the devs (or I) can't get in to actually fool around with this.  I will say that Vimeo content has been an area of intense work by the devs & things have improved for Vimeo content.  But it seems there's still some way to go before all the bugs are ironed out.  I can only hope @paulrouget can take what you've provided & recognize something he can fix.

If you're really into self-abuse, you can go back to our old forum https://groups.google.com/g/video-downloadhelper-q-and-a & hunt for a thread with this title:

Too fast video - possible solutions or workarounds.

It deals with a number of issues, too fast videos among them.  If you can apply any of what we discuss in there to your site, you might be able to download your content.  Failing that, the only thing I can suggest is to wait for Paul to peek in here & say something helpful.

**Comment by paulrouget** [comment]
@BundyBob thank you for the bug report!

I think I figured it out. I pushed a fix in the beta version (version 8.2.1.2).

Mind giving it a try? https://www.downloadhelper.net/beta

**Reply by BundyBob** [reply]
@paulrouget, thanks for the beta file.  I have some testing done - this time with Youtube sites rather than behind a Paywall so that you can test.  I'll attach an image of testing to date.  I rebooted the computer after a test but forgot to save the image at that point so what will be attached is about half way in testing but does show the context menu media list issue.  I was working on sending you a ZIP of the portable Firefox environment so that testing could be replicated but have now realised I would be sending my license code with it (oops) so I will need to rebuilt the test framework.  Do you have a separate licence code for testing that I can enter into the setup before passing it on to you guys?
A test not in this image is the next instance I relaunched the browser & the menu lists differed - same 2 tab windows but 2 Media lists for the 'orange' tab rather than the 'black' tab (ie the colour scheme I used in the image for ease of identification).
Cheers @BundyBob 


![Beta_8 2 0 23a1_Tests_#01](https://github.com/aclap-dev/video-downloadhelper/assets/111411485/8a2d0be7-79a4-4d56-ae97-0bdfb82b9bb8)

![Beta_8 2 0 23a1_Tests_#02](https://github.com/aclap-dev/video-downloadhelper/assets/111411485/5eb07c72-054f-4950-991e-19b6a73d166b)

**Reply by BundyBob** [reply]
@paulrouget , just to check, the beta file via the link shows as: Version 8.2.0.23a1 (beta)
I assumed this internal number for the beta of release 8.2.1.2 and corresponds with your comment "_I pushed a fix in the beta version (version 8.2.1.2)_"
Let me know if this is not the case.
Cheers, @BundyBob

**Reply by paulrouget** [reply]
@BundyBob if you go to https://www.downloadhelper.net/beta and install the **Firefox beta version**, it should install version `8.2.1.2a1`. Isn't that what you see? If not, go into your addon settings, and check if there's any update pending for the addon.

**Reply by BundyBob** [reply]
@paulrouget, OK not sure how I did it but finally it is loaded (as below).  A preliminary test looks promising but I will reset my test environment to confirm all is well.  I will be away for several hours but will get back onto it tonight & let you know (hope I can fluke my way to updating the add-on in that environment ;)
Cheers, 
@BundyBob 

Video DownloadHelper
Version 8.2.1.2a1 (beta)
Browser locale: en-US
Production build
Built on 2024-02-21
Build options: browser=firefox, linuxlic=false, noyt=false, target=mozilla
